devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+```
+
+`cancellation_token_create` creates a new cancellation token with the specified initial cancellation state.
+
+**SRS_CANCELLATION_TOKEN_04_001: [** `cancellation_token_create` shall allocate memory for a `THANDLE(CANCELLATION_TOKEN)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_002: [** If any underlying error occurs `cancellation_token_create` shall fail and return `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_003: [** `cancellation_token_create` shall create a `TCALL_DISPATCHER` handle by calling `TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_004: [** `cancellation_token_create` shall set the initial state to be equal to `canceled` parameter. **]**
+
+**SRS_CANCELLATION_TOKEN_04_005: [** `cancellation_token_create` shall return a valid `THANDLE(CANCELLATION_TOKEN)` when successful. **]**
+
+## cancellation_token_is_canceled
+
+```c
+MOCKABLE_FUNCTION(, int, cancellation_token_is_canceled, THANDLE(CANCELLATION_TOKEN), cancellation_token, bool*, canceled);
+```
+
+`cancellation_token_is_canceled` assigns `true` to `*canceled` if the cancellation token has been canceled and `false` otherwise.
+
+**SRS_CANCELLATION_TOKEN_04_006: [** `cancellation_token_is_canceled` shall return a non-zero value if `cancellation_token` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_007: [** `cancellation_token_is_canceled` shall return a non-zero value if `canceled` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_008: [** `cancellation_token_is_canceled` shall assign `true` to `*canceled` if the token has been canceled and `false` otherwise. **]**
+
+**SRS_CANCELLATION_TOKEN_04_009: [** `cancellation_token_is_canceled` shall return `0` when successful. **]**
+
+## cancellation_token_register_notify
+
+```c
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context);
+```
+
+`cancellation_token_register_notify` registers `on_cancel` to be called when the token is canceled.
+
+**SRS_CANCELLATION_TOKEN_04_010: [** `cancellation_token_register_notify` shall fail and return `NULL` when `cancellation_token` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_011: [** `cancellation_token_register_notify` shall fail and return `NULL` when `on_cancel` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_012: [** `cancellation_token_register_notify` shall fail and return `NULL` when any underlying call fails. **]**
+
+**SRS_CANCELLATION_TOKEN_04_013: [** `cancellation_token_register_notify` shall call `TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)` to register `on_cancel` with the dispatcher. **]**
+
+**SRS_CANCELLATION_TOKEN_04_014: [** `cancellation_token_register_notify` shall call `THANDLE_MALLOC(CANCELLATION_TOKEN_REGISTRATION)` to allocate a `CANCELLATION_TOKEN_REGISTRATION` handle. **]**
```

Comment:
>CANCELLATION_TOKEN_REGISTRATION [](http://example.com/codeflow?start=101&length=31)

What happens if between 04_013 and 04_014 `cancellation_token_cancel` is called AND 04_014 fails?

I suspect that `on_cancel` will be called, but the function return is NULL which would be "weird".

Can we do without `TCALL_DISPATCHER_UNREGISTER_TARGET`? Maybe switching 04_013 and 04_14?

...

Reply:
Good catch. Should have thought of that. Matt noted the same thing. Will fix.

Reply:
Fixed.